Sprint 14 solution

Dvoychenkov opened the pull request.

Dvoychenkov committed:
- Добавлен базовый набор Repository, Service, Controller для работы с пользователями;
- Добавлена модель, DTO и маппер для пользователя;
- Добавлены утилиты для валидации;
- Добавлено исключение;

Dvoychenkov committed:
- Добавлены исключения;
- Добавлена валидация;
- Скорректирована реализация процессов по пользователю с учётом коллекции постмана и правки багов;

Dvoychenkov committed:
- Добавлено всё по теме вещей согласно ТЗ;
- Добавлено необходимое исключение и его сопоставление с http-статусом для ответа;
- Внесены сопутствующие изменения и косметические правки в классы по теме функционала пользователя;

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -0,0 +1,39 @@
+        return item;
+    }
+
+    public static void updateItem(Item item, UpdateItemDto updateItemDto) {
```
С маппером все ок, но желательно использовать MapStruct

Dvoychenkov replied:
Спасибо, годный вариант, упростил код

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepositoryInMemory.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public Item save(Item item) {
+        if (item == null) return null;
```
Фигурные скобки нужно использовать всегда

Dvoychenkov replied:
Поправил везде, где нашёл

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepositoryInMemory.java:
```diff
@@ -0,0 +1,66 @@
+                    if (item.isAvailable() != isAvailable) return false;
+                    if (doNameSearch) {
+                        String itemName = item.getName();
+                        if (itemName != null && itemName.toLowerCase().contains(nameSubstring.toLowerCase())) {
```
Можно использовать метод containsIgnoreCase

Dvoychenkov replied:
Добавил его использование и немного отрефакторил

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceBase.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public ItemDto save(Long id, UpdateItemDto updateItemDto, Long ownerId) {
+        userService.getUserOrThrow(ownerId);
```
Если вы хотите только проверить, что сущность существует - можно использовать метод existsById, позволяет не выгружать сущность в память, а так же не выполнять доп запросы, которые могут быть если  в сущности есть связи ManyToOne и подобные

Dvoychenkov replied:
Добавил отдельный метод только для проверки без извлечения и внёс соответствующие правки по коду сервисов

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,10 @@
+
+import lombok.Value;
+
+@Value
```
Здорово, что используете эту аннотацию

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryInMemory.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public Collection<User> findAll() {
+        return users.values();
```
Нужно дополнительно обернуть в новую коллекцию

Dvoychenkov replied:
Обернул

Dvoychenkov committed:
- Упрощёны мапперы значений с использованием MapStruct;

Dvoychenkov committed:
- Добавлены недостающие фигурные скобки для условий согласно кодстайлу;

Dvoychenkov committed:
- Упрощён метод поиска вещей с использованием StringUtils из либы Apache Commons;

Dvoychenkov committed:
- Вынес формирование предиката для фильтрации вещей в отдельный метод;

Dvoychenkov committed:
- Добавил в утилиты валидации метод по проверке наличия и выбросу эксепшна при отсутствии;
- Репозиторий пользователя дополнен методом проверки наличия объекта по ключу без извлечения;
- Сервисы пользователя и вещей теперь не получают объект, если нужно только проверить его наличие;

Dvoychenkov committed:
- Добавлено оборачивание в новую неизменяемую коллекцию результатов поиска юзеров;

VladimirTitov4 commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -0,0 +1,20 @@
+import ru.practicum.shareit.item.model.Item;
+
+@Mapper(componentModel = "spring")
+public interface ItemMapper {
```
Здорово. что получилось применить мапстракт !

Вместо строки "spring" можно использовать константу 
org.mapstruct.MappingConstants.ComponentModel.SPRING

Dvoychenkov committed:
- Поправлено настройка компонентной модели для мапперов на использование контстанты;
